### PR TITLE
feat: #158 - introduction to folksonomy

### PR DESCRIPTION
**Pull request by monsieurtanuki**
New files:
* `api_folksonomy_test.dart`: Tests around folksonomy
* `folksonomy.dart`: Client calls of the Folksonomy API (Open Food Facts)
* `KeyStats.dart`: Folksonomy: statistics around a tag key.
* `ProductList.dart`: Folksonomy: current value for a product and a tag key.
* `ProductStats.dart`: Folksonomy: statistics about the tag keys on a product.
* `ProductTag.dart`: Folksonomy product tag: for this barcode, that value is set for that key.

Impacted files:
* `HttpHelper.dart`: added "delete" and "put" requests
* `JsonHelper.dart`: added a helper method around timestamps
* `OpenFoodAPIConfiguration.dart`: added reference URIs for folksonomy
* `openfoodfacts.dart`: exported folksonomy files
* `UriHelper.dart`: added a helper method for folksonomy URIs

### CODE REVIEW

**Review by M123-dev — APPROVED** [review]
Thanks @monsieurtanuki, the code looks really good. I have left a few comments but feel free to ignore them as they are almost all about comments.

File: lib/folksonomy.dart [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+import 'utils/UriHelper.dart';
+
+/// Client calls of the Folksonomy API (Open Food Facts)
+///
+/// cf. https://api.folksonomy.openfoodfacts.org/docs
+class FolksonomyAPIClient {
+  FolksonomyAPIClient._();
+
+  /// "hello world"
+  static Future<void> hello({
```
[the remark this answers is not in the document]
Comment (monsieurtanuki) [reply lib/folksonomy.dart:21]: As for `ping`, returning a `void` does make sense. At least for me, and that's an interesting point.

If you compare to the current code in openfoodfacts, for instance https://github.com/openfoodfacts/openfoodfacts-dart/blob/master/lib/openfoodfacts.dart#L886-L890, we see that @jasmeet0817 has a problem with throwing an `Exception`. Should the methods handle the exception and return a clean useless value (and a step further: call Sentry) or let the developer handle the exception?

I think we should let the exceptions being thrown. We should tell the developers that our methods throw exceptions (e.g. when there's no internet, or when the server is down), and let them handle the exceptions their own way, for instance with Sentry in smoothie.

In the specific case of `hello` or `ping`, we don't give a damn about a value: we just want to know if everything was alright.

**Comment by monsieurtanuki** [comment]
Thank you @M123-dev for your code review!
I tried to be more explicit in the comments, and I added a couple of tests.
Still don't know if that makes sense to embed folksonomy here, but we can think about it until we publish the package.

**Comment by M123-dev** [comment]
@monsieurtanuki, lets do a new release when #336 is merged, mainly because I need the new globalUser for https://github.com/openfoodfacts/smooth-app/issues/700 but we can either remove the folksonomy from the CHANGELOG (and introduce it when its finsihed) or add to it, that its in work and may change.

**Comment by monsieurtanuki** [comment]
@M123-dev Do what is convenient for you. folksonomy is not a top priority, smoothie is.
Beyond that, I'm not a git expert and I don't know how you can remove folksonomy from the release or the CHANGELOG, and how safe this operation would be. I'm glad there are bold people like you to do this :)

**Comment by M123-dev** [comment]
@monsieurtanuki 


Charles answered:

> There is no test server for the moment. Meanwhile there is a complete doc how to install a test version of the API: https://github.com/openfoodfacts/folksonomy_api/blob/main/INSTALL.md


> I think I will install a test server, but at the beginning of 2022. If it's very important I can try to install it before, tell me.

**Comment by monsieurtanuki** [comment]
Thank you @M123-dev and Charles!
Actually I don't need a test server, I only need a test user for integration tests. That could be on PROD env.

**Pull request merged**